blockchain/src/block/read.rs

```diff
@@ -17,14 +20,58 @@ pub trait BlockQueries: Debug {
         include_transactions: bool,
     ) -> Result<Option<BlockResponse>, Self::Err>;
 
-    fn by_height(
+    fn get_forkchoice_state(&self, storage: &Self::Storage) -> Result<ForkchoiceState, Self::Err>;
+
+    fn height_to_hash(&self, storage: &Self::Storage, height: u64) -> Result<B256, Self::Err>;
+
+    /// This function returns `true` if the given block hash `maybe_ancestor` is reachable
+    /// via parent hash links in the blockchain starting from the block hash `head` or
+    /// `maybe_ancestor == head`.
+    /// Note: the default implementation is rather inefficient (it always walks the whole chain
+    /// from `head` to `maybe_ancestor`); implementors of this trait should provide a more streamlined
+    /// implementation (e.g. using a skip list).
```

Comment:
I've looked up how (op) geth does it, and apparently instead of walking the chain they try to do a cleaner separation of side and canonical chains. This means they always look up the head block in FCU by hash and height from header, and a mismatch signals that the canonical chain has changed, which triggers a reorg. IDK if that would be a better design but still.